File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/delius/model/DeliusUserPersonDetails.java
package uk.gov.justice.digital.hmpps.oauth2server.delius.model;

import lombok.Builder;
import lombok.Data;
import org.springframework.security.core.GrantedAuthority;
import uk.gov.justice.digital.hmpps.oauth2server.security.UserPersonDetails;

import java.util.Collection;

@Builder
@Data
public class DeliusUserPersonDetails implements UserPersonDetails {
    private String surname;
    private String firstName;
    private String username;
    private boolean locked;
    private Collection<? extends GrantedAuthority> roles;

    @Override
    public String getUserId() {
        return username;
    }

    @Override
    public String getName() {
        return String.format("%s %s", firstName, surname);
    }

    @Override
    public boolean isAdmin() {
        return false;
    }

    @Override
    public String getAuthSource() {
        return "delius";
    }

    @Override
    public void eraseCredentials() {
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return roles;
    }

    @Override
    public String getPassword() {
        return "password";
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return !locked;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return true;
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/delius/model/DeliusUserPersonDetails.kt
package uk.gov.justice.digital.hmpps.oauth2server.delius.model

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User
import uk.gov.justice.digital.hmpps.oauth2server.security.AuthSource
import uk.gov.justice.digital.hmpps.oauth2server.security.UserPersonDetails

data class DeliusUserPersonDetails(
  private val username: String,
  override val userId: String,
  override val firstName: String,
  private val surname: String,
  val email: String,
  private val enabled: Boolean = false,
  private val locked: Boolean = false,
  private val roles: Collection<GrantedAuthority?> = emptySet(),
) : UserPersonDetails {

  override fun getUsername(): String = username

  override val name: String
    get() = "$firstName $surname"

  override val isAdmin: Boolean = false

  override val authSource: String
    get() = "delius"

  override fun toUser(): User =
    User(
      username = username,
      source = AuthSource.delius,
      email = email,
      verified = true,
      enabled = enabled,
    )

  override fun eraseCredentials() {}

  // add in ROLE_PROBATION to standard roles
  override fun getAuthorities(): Collection<GrantedAuthority?> = roles.plus(SimpleGrantedAuthority("ROLE_PROBATION"))

  override fun getPassword(): String = "password"

  override fun isAccountNonExpired(): Boolean = true

  override fun isAccountNonLocked(): Boolean = !locked

  override fun isCredentialsNonExpired(): Boolean = true

  override fun isEnabled(): Boolean = enabled
}
